circuit-lib/circuitlib-rs/src/gnark/helpers.rs

```diff
@@ -11,46 +11,65 @@ use serde_json::json;
 
 use crate::gnark::constants::{HEALTH_CHECK, SERVER_ADDRESS};
 
-pub fn spawn_gnark_server(path: &str, wait_time: u64) -> Child {
-    let server_process = Command::new("sh")
-        .arg("-c")
-        .arg(path)
-        .spawn()
-        .expect("Failed to start server process");
-
-    // Wait for the server to launch before proceeding.
-    thread::sleep(Duration::from_secs(wait_time));
+static IS_LOADING: AtomicBool = AtomicBool::new(false);
 
-    server_process
+pub async fn spawn_gnark_server(
+    path: &str,
+    restart: bool,
+    inclusion: bool,
+    non_inclusion: bool,
+    combined: bool,
+) {
+    if restart {
+        kill_gnark_server();
+    }
+    if !health_check(1, 3).await && !IS_LOADING.load(Ordering::Relaxed) {
+        IS_LOADING.store(true, Ordering::Relaxed);
+        let mut arg: String = "".to_string();
+        if inclusion {
+            arg += "inclusion";
+        } else if non_inclusion {
+            arg += "non-inclusion";
+        } else if combined {
+            arg += "combined";
+        }
+        Command::new("sh")
+            .arg("-c")
+            .arg(format!("{} {}", path, arg))
+            .spawn()
+            .expect("Failed to start server process");
+        health_check(20, 5).await;
+        IS_LOADING.store(false, Ordering::Relaxed);
+    }
 }
 
-pub fn kill_gnark_server(gnark: &mut Child) {
+pub fn kill_gnark_server() {
     Command::new("sh")
         .arg("-c")
         .arg("killall light-prover")
         .spawn()
         .unwrap();
-    gnark.kill().unwrap();
```

Comment:
Could we try to avoid calling `sh -c killall` and kill the processes with native Rust?

Like, for example

```rust
use sysinfo::{System, SystemExt, ProcessExt, Signal};

pub fn kill_gnark_server() -> Result<()> {
    let mut system = System::new_all();
    system.refresh_all();

    for (pid, process) in system.processes() {
        if process.name() == "light-prover" {
            process.kill(Signal::Term)?;
        }
    }
}
```